File: benchmarks/workloads/trees/trees.go
// Binary trees: allocation-heavy recursive tree build and check, depth 18.
// Timed region: everything. Checksum: sum of all node counts.
// Go frees nothing explicitly; this workload measures its garbage collector.
package main

import (
	"fmt"
	"time"
)

const (
	minDepth = 4
	maxDepth = 18
)

type Node struct{ left, right *Node }

func bottomUp(depth int) *Node {
	if depth == 0 {
		return &Node{}
	}
	return &Node{bottomUp(depth - 1), bottomUp(depth - 1)}
}

func check(node *Node) int64 {
	var left, right int64
	if node.left != nil {
		left = check(node.left)
	}
	if node.right != nil {
		right = check(node.right)
	}
	return 1 + left + right
}

func main() {
	start := time.Now()
	var total int64
	stretch := bottomUp(maxDepth + 1)
	total += check(stretch)
	longLived := bottomUp(maxDepth)
	for depth := minDepth; depth <= maxDepth; depth += 2 {
		iterations := 1 << (maxDepth - depth + minDepth)
		for i := 0; i < iterations; i++ {
			total += check(bottomUp(depth))
		}
	}
	total += check(longLived)
	elapsedMs := float64(time.Since(start).Nanoseconds()) / 1e6
	fmt.Printf("max_depth %d\n", maxDepth)
	fmt.Printf("elapsed_ms %.3f\n", elapsedMs)
	fmt.Printf("checksum %d\n", total)
}

File: benchmarks/workloads/trees/trees.rs
// Binary trees: allocation-heavy recursive tree build and check, depth 18.
// Timed region: everything. Checksum: sum of all node counts.
const MIN_DEPTH: u32 = 4;
const MAX_DEPTH: u32 = 18;

struct Node { left: Option<Box<Node>>, right: Option<Box<Node>> }

fn bottom_up(depth: u32) -> Box<Node> {
    if depth == 0 {
        Box::new(Node { left: None, right: None })
    } else {
        Box::new(Node { left: Some(bottom_up(depth - 1)), right: Some(bottom_up(depth - 1)) })
    }
}

fn check(node: &Node) -> i64 {
    let left = node.left.as_ref().map_or(0, |c| check(c));
    let right = node.right.as_ref().map_or(0, |c| check(c));
    1 + left + right
}

fn main() {
    let start = std::time::Instant::now();
    let mut total: i64 = 0;
    let stretch = bottom_up(MAX_DEPTH + 1);
    total += check(&stretch);
    drop(stretch);
    let long_lived = bottom_up(MAX_DEPTH);
    let mut depth = MIN_DEPTH;
    while depth <= MAX_DEPTH {
        let iterations = 1u32 << (MAX_DEPTH - depth + MIN_DEPTH);
        for _ in 0..iterations {
            let tree = bottom_up(depth);
            total += check(&tree);
        }
        depth += 2;
    }
    total += check(&long_lived);
    let elapsed_ms = start.elapsed().as_secs_f64() * 1000.0;
    println!("max_depth {}", MAX_DEPTH);
    println!("elapsed_ms {:.3}", elapsed_ms);
    println!("checksum {}", total);
}
